### ISSUE TITLE
Architecture ppc64le , mas-mongo-ce-1 & mas-mongo-ce-2 statefulset reboots once in a day

### ISSUE THREAD

**Issue by snehlatamohite** [mongodb/mongodb-kubernetes-operator#1013]
Hi All,

We build operator images locally on ppc64le and try installing operator on OCP 4.10 cluster. We are using MongoDB for development purpose only.  We observed that mas-mongo-ce-1   and mas-mongo-ce-1  reboots once in a day.  Any help or pointers to debug issue further and fix it will be helpful. 

```
NAME                                           READY   STATUS        RESTARTS        AGE
mas-mongo-ce-0                                 2/2     Running       0               28d
mas-mongo-ce-1                                 2/2     Running       64 (148m ago)   28d
mas-mongo-ce-2                                 2/2     Running       55 (3d6h ago)   28d
mongodb-kubernetes-operator-6dbb5fd44d-kbjv7   1/1     Running       1               28d

```

**Comment by slaskawi** [comment]
In order to debug this issue, we need to know why the container was rebooted. 

The easiest way to do it is to check `kubectl describe pod mas-mongo-ce-1`. You may also try to diagnose the StatefulSet itself `kubectl describe statefulset mas-mongo-ce`.

You may find more information on troubleshooting on the internet: https://damieng.com/blog/2015/04/24/make-home-end-keys-behave-like-windows-on-mac-os-x/

**Comment by snehlatamohite** [comment]
@slaskawi  - Thanks for pointer. Actually no errors as such in statefulsets. 
```
Name:               mas-mongo-ce
Namespace:          mongotest
CreationTimestamp:  Wed, 18 May 2022 05:36:50 -0400
Selector:           app=mas-mongo-ce-svc
Labels:             app=mas-mongo-ce-svc
Annotations:        <none>
Replicas:           3 desired | 3 total
Update Strategy:    RollingUpdate
Pods Status:        3 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:           app=mas-mongo-ce-svc
  Service Account:  mongodb-kubernetes-operator
  Init Containers:
   mongod-posthook:
    Image:      wiotp-docker-local.artifactory.swg-devops.com/cpopen/staging/ppc/mongodb-ce-operator-images/v0.7.0/community-operator-version-upgrade-post-start-hook
    Port:       <none>
    Host Port:  <none>
    Command:
      cp
      version-upgrade-hook
      /hooks/version-upgrade
    Environment:  <none>
    Mounts:
      /hooks from hooks (rw)
   mongodb-agent-readinessprobe:
    Image:      wiotp-docker-local.artifactory.swg-devops.com/cpopen/staging/ppc/mongodb-ce-operator-images/v0.7.0/mongodb-kubernetes-readiness
    Port:       <none>
    Host Port:  <none>
    Command:
      cp
      /probes/readinessprobe
      /opt/scripts/readinessprobe
    Environment:  <none>
    Mounts:
      /opt/scripts from agent-scripts (rw)
  Containers:
   mongod:
    Image:      wiotp-docker-local.artifactory.swg-devops.com/cpopen/staging/ppc/mongodb/mongo:4.2.6-focal
    Port:       <none>
    Host Port:  <none>
    Command:
      /bin/sh
      -c

      #run post-start hook to handle version changes
      /hooks/version-upgrade

      # wait for config and keyfile to be created by the agent
       while ! [ -f /data/automation-mongod.conf -a -f /var/lib/mongodb-mms-automation/authentication/keyfile ]; do sleep 3 ; done ; sleep 2 ;

      # with mongod configured to append logs, we need to provide them to stdout as
      # mongod does not write to stdout and a log file
      tail -F /var/log/mongodb-mms-automation/mongodb.log > /dev/stdout &

      # start mongod with this configuration
      exec mongod -f /data/automation-mongod.conf;


    Limits:
      cpu:     1
      memory:  1G
    Requests:
      cpu:     500m
      memory:  400M
    Environment:
      AGENT_STATUS_FILEPATH:  /healthstatus/agent-health-status.json
    Mounts:
      /data from data-volume (rw)
      /healthstatus from healthstatus (rw)
      /hooks from hooks (rw)
      /var/lib/mongodb-mms-automation/authentication from mas-mongo-ce-keyfile (rw)
      /var/lib/tls/ca/ from tls-ca (ro)
      /var/lib/tls/server/ from tls-secret (ro)
      /var/log/mongodb-mms-automation from logs-volume (rw)
   mongodb-agent:
    Image:      wiotp-docker-local.artifactory.swg-devops.com/cpopen/staging/ppc/mongodb-ce-operator-images/v0.7.0/mongodb-agent-dev
    Port:       <none>
    Host Port:  <none>
    Command:
      /bin/bash
      -c
      current_uid=$(id -u)
      declare -r current_uid
      if ! grep -q "${current_uid}" /etc/passwd ; then
      sed -e "s/^mongodb:/builder:/" /etc/passwd > /tmp/passwd
      echo "mongodb:x:$(id -u):$(id -g):,,,:/:/bin/bash" >> /tmp/passwd
      export NSS_WRAPPER_PASSWD=/tmp/passwd
      export LD_PRELOAD=libnss_wrapper.so
      export NSS_WRAPPER_GROUP=/etc/group
      fi
      agent/mongodb-agent -cluster=/var/lib/automation/config/cluster-config.json -healthCheckFilePath=/var/log/mongodb-mms-automation/healthstatus/agent-health-status.json -serveStatusPort=5000 -skipMongoStart -noDaemonize -useLocalMongoDbTools
    Limits:
      cpu:     1
      memory:  500M
    Requests:
      cpu:      500m
      memory:   400M
    Readiness:  exec [/opt/scripts/readinessprobe] delay=5s timeout=1s period=10s #success=1 #failure=40
    Environment:
      AGENT_STATUS_FILEPATH:  /var/log/mongodb-mms-automation/healthstatus/agent-health-status.json
      AUTOMATION_CONFIG_MAP:  mas-mongo-ce-config
      HEADLESS_AGENT:         true
      POD_NAMESPACE:           (v1:metadata.namespace)
    Mounts:
      /data from data-volume (rw)
      /opt/scripts from agent-scripts (rw)
      /var/lib/automation/config from automation-config (ro)
      /var/lib/mongodb-mms-automation/authentication from mas-mongo-ce-keyfile (rw)
      /var/lib/tls/ca/ from tls-ca (ro)
      /var/lib/tls/server/ from tls-secret (ro)
      /var/log/mongodb-mms-automation from logs-volume (rw)
      /var/log/mongodb-mms-automation/healthstatus from healthstatus (rw)
  Volumes:
   agent-scripts:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
   automation-config:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  mas-mongo-ce-config
    Optional:    false
   healthstatus:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
   hooks:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
   mas-mongo-ce-keyfile:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
   tls-ca:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      mas-mongo-ce-cert-map
    Optional:  false
   tls-secret:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  mas-mongo-ce-server-certificate-key
    Optional:    false
Volume Claims:
  Name:          data-volume
  StorageClass:  nfs-storage-provisioner
  Labels:        <none>
  Annotations:   <none>
  Capacity:      5Gi
  Access Modes:  [ReadWriteOnce]
  Name:          logs-volume
  StorageClass:  nfs-storage-provisioner
  Labels:        <none>
  Annotations:   <none>
  Capacity:      2Gi
  Access Modes:  [ReadWriteOnce]
Events:          <none>

```

**Comment by slaskawi** [comment]
Yes, the events (which I'm the most interested in) are empty.

The default retention period of Kubernetes events is 1h. Could you please try to issue this command right after such a restart happens? Without this data, we won't be able to help you much.

**Comment by snehlatamohite** [comment]
@slaskawi  - Yes sure I will try to share details.

**Comment by snehlatamohite** [comment]
@slaskawi  - I observed below errors in mongo-agent container 
```
Now using project "mongo" on server "https://api.snehlatam-410-ocp1.ibm.com:6443".
[root@snehlatam-410-ocp1-mon01-bastion-0 ~]# oc logs mas-mongo-ce-1 -c mongodb-agent | grep .error
[2022-06-21T22:07:13.629+0000] [.error] [src/mongoclientservice/mongoclientservice.go:sendClientToRequesters:719] [22:07:13.629] Error getting client ready for conn params = mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017 (local=false).  Informing all requests and disposing of client (0xc0009f4a80).  requests=[ 0xc0009784e0 ] : connection(mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017[-8013]) incomplete read of message header: EOF
[2022-06-21T22:08:17.076+0000] [.error] [src/director/director.go:executePlan:923] <mas-mongo-ce-1> [22:08:17.076] <mas-mongo-ce-1> [22:08:17.076] Not running step 0(WaitRsInit) of move 1(WaitRsInit) because the preconditions aren't met:
[2022-06-21T22:23:27.472+0000] [.error] [src/mongoclientservice/mongoclientservice.go:sendClientToRequesters:719] [22:23:27.472] Error getting client ready for conn params = mas-mongo-ce-2.mas-mongo-ce-svc.mongo.svc.cluster.local:27017 (local=false).  Informing all requests and disposing of client (0xc00091d500).  requests=[ 0xc00002c6c0 ] : context deadline exceeded
[2022-06-22T09:03:18.252+0000] [.error] [src/mongoclientservice/mongoclientservice.go:sendClientToRequesters:719] [09:03:18.252] Error getting client ready for conn params = mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017 (local=false).  Informing all requests and disposing of client (0xc00078ee00).  requests=[ 0xc0006766c0 ] : connection(mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017[-15974]) incomplete read of message header: EOF
[root@snehlatam-410-ocp1-mon01-bastion-0 ~]# oc logs mas-mongo-ce-2 -c mongodb-agent | grep .error
[2022-06-21T22:07:15.318+0000] [.error] [src/mongoclientservice/mongoclientservice.go:sendClientToRequesters:719] [22:07:15.318] Error getting client ready for conn params = mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017 (local=false).  Informing all requests and disposing of client (0xc000176700).  requests=[ 0xc0005f8d20 ] : connection(mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017[-7725]) incomplete read of message header: EOF
[2022-06-21T22:23:16.976+0000] [.error] [src/mongoclientservice/mongoclientservice.go:sendClientToRequesters:719] [22:23:16.976] Error getting client ready for conn params = mas-mongo-ce-2.mas-mongo-ce-svc.mongo.svc.cluster.local:27017 (local=false).  Informing all requests and disposing of client (0xc000257500).  requests=[ 0xc0002cb8c0 ] : connection(mas-mongo-ce-2.mas-mongo-ce-svc.mongo.svc.cluster.local:27017[-8089]) incomplete read of message header: EOF
[2022-06-21T22:24:16.766+0000] [.error] [src/director/director.go:executePlan:923] <mas-mongo-ce-2> [22:24:16.766] <mas-mongo-ce-2> [22:24:16.766] Not running step 0(WaitRsInit) of move 1(WaitRsInit) because the preconditions aren't met:
[2022-06-22T09:03:24.771+0000] [.error] [src/mongoclientservice/mongoclientservice.go:sendClientToRequesters:719] [09:03:24.771] Error getting client ready for conn params = mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017 (local=false).  Informing all requests and disposing of client (0xc00095a380).  requests=[ 0xc0007f57a0 ] : connection(mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017[-15654]) incomplete read of message header: EOF
[2022-06-22T09:34:16.928+0000] [.error] [src/mongoclientservice/mongoclientservice.go:sendClientToRequesters:719] [09:34:16.928] Error getting client ready for conn params = mas-mongo-ce-2.mas-mongo-ce-svc.mongo.svc.cluster.local:27017 (local=false).  Informing all requests and disposing of client (0xc0007ac700).  requests=[ 0xc0007b9b00 ] : connection(mas-mongo-ce-2.mas-mongo-ce-svc.mongo.svc.cluster.local:27017[-16210]) incomplete read of message header: EOF
[2022-06-22T09:35:23.686+0000] [.error] [src/director/director.go:executePlan:923] <mas-mongo-ce-2> [09:35:23.686] <mas-mongo-ce-2> [09:35:23.686] Not running step 0(WaitRsInit) of move 1(WaitRsInit) because the preconditions aren't met:
[2022-06-22T19:49:33.947+0000] [.error] [src/mongoclientservice/mongoclientservice.go:sendClientToRequesters:719] [19:49:33.947] Error getting client ready for conn params = mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017 (local=false).  Informing all requests and disposing of client (0xc000504700).  requests=[ 0xc00072cf00 ] : context deadline exceeded

```

**Comment by slaskawi** [comment]
Thanks for the input @snehlatamohite. The errors that you see are most likely caused by the connectivity issues (EOFs, deadline exceeded, which translates into a timeout). If those errors happened in a system that was running for a while, I'd suspect network and firewalls. 

Could you please paste your CR content? Perhaps there's something suspicious there.

**Comment by snehlatamohite** [comment]
@slaskawi  - Thanks a lot for your response.  Please find below CR output. 
```

[root@snehlatam-410-ocp1-mon01-bastion-0 ~]# oc get MongoDBCommunity -n mongo -o yaml
apiVersion: v1
items:
- apiVersion: mongodbcommunity.mongodb.com/v1
  kind: MongoDBCommunity
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"mongodbcommunity.mongodb.com/v1","kind":"MongoDBCommunity","metadata":{"annotations":{},"name":"mas-mongo-ce","namespace":"mongo"},"spec":{"additionalMongodConfig":{"net.tls.allowInvalidCertificates":true,"net.tls.allowInvalidHostnames":true,"storage.wiredTiger.engineConfig.journalCompressor":"snappy"},"members":3,"security":{"authentication":{"modes":["SCRAM-SHA-1","SCRAM-SHA-256"]},"tls":{"caConfigMapRef":{"name":"mas-mongo-ce-cert-map"},"certificateKeySecretRef":{"name":"mas-mongo-ce-cert-secret"},"enabled":true}},"statefulSet":{"spec":{"selector":{},"serviceName":"mas-mongo-ce-svc","template":{"spec":{"containers":[{"name":"mongod","resources":{"limits":{"cpu":"1","memory":"1G"}}}]}},"volumeClaimTemplates":[{"metadata":{"name":"data-volume"},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"5Gi"}},"storageClassName":"nfs-storage-provisioner"}},{"metadata":{"name":"logs-volume"},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"2Gi"}},"storageClassName":"nfs-storage-provisioner"}}]}},"type":"ReplicaSet","users":[{"db":"admin","name":"admin","passwordSecretRef":{"name":"mas-mongo-ce-admin-password"},"roles":[{"db":"admin","name":"clusterAdmin"},{"db":"admin","name":"userAdminAnyDatabase"},{"db":"admin","name":"dbOwner"},{"db":"admin","name":"readWriteAnyDatabase"}],"scramCredentialsSecretName":"mas-mongo-ce-scram"}],"version":"4.2.6-focal"}}
      mongodb.com/v1.lastAppliedMongoDBVersion: 4.2.6-focal
      mongodb.com/v1.lastSuccessfulConfiguration: '{"members":3,"type":"ReplicaSet","version":"4.2.6-focal","arbiters":0,"security":{"authentication":{"modes":["SCRAM-SHA-1","SCRAM-SHA-256"],"ignoreUnknownUsers":true},"tls":{"enabled":true,"optional":false,"certificateKeySecretRef":{"name":"mas-mongo-ce-cert-secret"},"caConfigMapRef":{"name":"mas-mongo-ce-cert-map"}}},"users":[{"name":"admin","db":"admin","passwordSecretRef":{"name":"mas-mongo-ce-admin-password","key":""},"roles":[{"db":"admin","name":"clusterAdmin"},{"db":"admin","name":"userAdminAnyDatabase"},{"db":"admin","name":"dbOwner"},{"db":"admin","name":"readWriteAnyDatabase"}],"scramCredentialsSecretName":"mas-mongo-ce-scram"}],"statefulSet":{"spec":{}},"additionalMongodConfig":{}}'
    creationTimestamp: "2022-06-21T11:15:05Z"
    generation: 1
    name: mas-mongo-ce
    namespace: mongo
    resourceVersion: "77537"
    uid: 765b6e2a-850b-4973-989e-b4225c86b137
  spec:
    additionalMongodConfig:
      net.tls.allowInvalidCertificates: true
      net.tls.allowInvalidHostnames: true
      storage.wiredTiger.engineConfig.journalCompressor: snappy
    members: 3
    security:
      authentication:
        ignoreUnknownUsers: true
        modes:
        - SCRAM-SHA-1
        - SCRAM-SHA-256
      tls:
        caConfigMapRef:
          name: mas-mongo-ce-cert-map
        certificateKeySecretRef:
          name: mas-mongo-ce-cert-secret
        enabled: true
    statefulSet:
      spec:
        selector: {}
        serviceName: mas-mongo-ce-svc
        template:
          spec:
            containers:
            - name: mongod
              resources:
                limits:
                  cpu: "1"
                  memory: 1G
        volumeClaimTemplates:
        - metadata:
            name: data-volume
          spec:
            accessModes:
            - ReadWriteOnce
            resources:
              requests:
                storage: 5Gi
            storageClassName: nfs-storage-provisioner
        - metadata:
            name: logs-volume
          spec:
            accessModes:
            - ReadWriteOnce
            resources:
              requests:
                storage: 2Gi
            storageClassName: nfs-storage-provisioner
    type: ReplicaSet
    users:
    - db: admin
      name: admin
      passwordSecretRef:
        name: mas-mongo-ce-admin-password
      roles:
      - db: admin
        name: clusterAdmin
      - db: admin
        name: userAdminAnyDatabase
      - db: admin
        name: dbOwner
      - db: admin
        name: readWriteAnyDatabase
      scramCredentialsSecretName: mas-mongo-ce-scram
    version: 4.2.6-focal
  status:
    currentMongoDBMembers: 3
    currentStatefulSetReplicas: 3
    mongoUri: mongodb://mas-mongo-ce-0.mas-mongo-ce-svc.mongo.svc.cluster.local:27017,mas-mongo-ce-1.mas-mongo-ce-svc.mongo.svc.cluster.local:27017,mas-mongo-ce-2.mas-mongo-ce-svc.mongo.svc.cluster.local:27017
    phase: Running
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

```

**Comment by slaskawi** [comment]
@snehlatamohite Nothing suspicious there but could I ask you to verify one thing - what is the expiration time you set on the certificates? Does any certificate renewal happen when the pods are being restarted?

**Comment by snehlatamohite** [comment]
@slaskawi  - I will check on  certificate and if certificate renewal happen when the pods are being restarted and then will revert.

**Comment by snehlatamohite** [comment]
@slaskawi  - Thanks for your support and guidance. I build operator and needed images and pushed to local artifactory and  replcaed images presebt in https://github.com/ibm-watson-iot/iot-docs/blob/master/mongodb/config/manager/__manager__.yaml with ppc64le images. I used below script for installing mongodb community operator https://github.com/ibm-watson-iot/iot-docs/blob/master/mongodb/install-mongo-ce.sh

Certs created using this script below script where in expiry is -days 3650 
https://github.com/ibm-watson-iot/iot-docs/blob/master/mongodb/certs/generateSelfSignedCert.sh#L11

**Comment by slaskawi** [comment]
@snehlatamohite So that's not it. This routes us back to the events. Please try to find out what was the cause of a restart.

**Comment by snehlatamohite** [comment]
@slaskawi  - I started a cron job on cluster which runs "oc describe statefulset mas-mongo-ce | grep Events >> logs.txt" every 15 minute.  But I see below in events 
```

Events:          <none>
Events:          <none>
Events:          <none>
Events:          <none>
```

And below events in restarted pod - 
oc describe pod mas-mongo-ce-2
```

Events:
  Type     Reason     Age                 From     Message
  ----     ------     ----                ----     -------
  Normal   Created    78m (x78 over 35d)  kubelet  Created container mongod
  Normal   Started    78m (x78 over 35d)  kubelet  Started container mongod
  Normal   Pulled     78m (x77 over 35d)  kubelet  Container image "wiotp-docker-local.artifactory.swg-devops.com/cpopen/staging/ppc/mongodb/mongo:4.2.6-focal" already present on machine
  Warning  Unhealthy  77m (x60 over 35d)  kubelet  Readiness probe failed:

```

```
oc get pods
NAME                                           READY   STATUS    RESTARTS       AGE
mas-mongo-ce-0                                 2/2     Running   0              35d
mas-mongo-ce-1                                 2/2     Running   77 (9h ago)    35d
mas-mongo-ce-2                                 2/2     Running   77 (79m ago)   35d
```

**Comment by dan-mckean** [comment]
Hi, I'm Dan and I'm the Product Manager for MongoDB's support of Kubernetes.

I'm doing some work right now to try and identify how the Community Operator is being used. The Community Operator is something I inherited when I started at MongoDB, but it doesn't get as much attention from us as we'd like and we're trying to understand how it's used in order to establish its future. It will help us prioritize future issues and PRs raised by the community 🙂

Here's a super short survey (it's much easier for us to review all the feedback that way!): [https://docs.google.com/forms/d/e/1FAIpQLSfwrwyxBSlUyJ6AmC-eYlgW_3JEdfA48SB2i5--_WpiynMW2w/viewform?usp=sf_link](https://docs.google.com/forms/d/e/1FAIpQLSfwrwyxBSlUyJ6AmC-eYlgW_3JEdfA48SB2i5--_WpiynMW2w/viewform?usp=sf_link)

 If you'd rather email me instead: [dan.mckean@mongodb.com](mailto:dan.mckean@mongodb.com?subject=MongoDB%20Community%20Operator%20feedback)

Thank you in advance!
Dan
